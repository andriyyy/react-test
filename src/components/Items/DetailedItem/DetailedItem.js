import React, { Component } from "react";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import { compose } from "recompose";
import Moment from "react-moment";
import {
  List,
  Paper,
  withStyles,
  ListItem,
  Typography,
  Button,
} from "@material-ui/core";
import IconButton from "@material-ui/core/IconButton";
import KeyboardBackspace from "@material-ui/icons/KeyboardBackspace";
import Link from "@material-ui/core/Link";
import {
  getId,
  getAttendeesIds,
  getAttendeesHasErrored,
  getAttendeesIsLoading,
  getItem,
  getItems,
  getUsersKey,
  getAttendeeFormatted,
  getAuthUser,
} from "../../../selectors/Selectors";
import { attendeesIdsFetchData } from "../../../actions/users";
import {
  itemsOff,
  usersOff,
  usersEnrolmentsListOff,
  reject,
  notReject,
  updateItemsInState,
} from "../../../actions/firebase";

const styles = (theme) => ({
  margin: {
    margin: theme.spacing.unit * 2,
  },
  padding: {
    padding: theme.spacing.unit,
  },
  main: {
    width: "auto",
    display: "block", // Fix IE 11 issue.
    marginLeft: theme.spacing.unit * 3,
    marginRight: theme.spacing.unit * 3,
    [theme.breakpoints.up(400 + theme.spacing.unit * 3 * 2)]: {
      minWidth: 120,
      maxWidth: 450,
      marginLeft: "auto",
      marginRight: "auto",
    },
  },
  paper: {
    marginTop: theme.spacing.unit * 8,
    display: "flex",
    flexDirection: "column",
    alignItems: "center",
    padding: `${theme.spacing.unit * 2}px ${theme.spacing.unit * 3}px ${
      theme.spacing.unit * 3
    }px`,
  },
  submit: {
    marginTop: theme.spacing.unit * 3,
  },
  formControl: {
    margin: theme.spacing.unit,
    minWidth: 120,
    maxWidth: 300,
  },
  image: {
    maxWidth: 250,
  },
  button: {
    marginRight: 8,
  },
});

class DetailedItem extends Component {
  constructor(props) {
    super(props);
    this.state = {
      active: false,
    };
  }
  componentDidMount() {
    const id = this.props.getId();
    this.props.fetchAttendeesIds(id);
  }

  componentWillUnmount() {
    this.props.onItemsOff();
    this.props.onUsersOff();
    this.props.onUsersEnrolmentsListOff();
  }

  onRedirect = (event) => {
    event.preventDefault();
    this.props.history.push(`/home`);
  };

  onView = (id) => {
    this.props.history.push(`/user/${id}`);
  };

  onReject = (uid, iid) => {
    this.props.onReject(uid, iid, this.addActiveToState);
  };

  addActiveToState = () => {
    this.setState({ active: true });
    this.props.onDeleteAttendee();
  };

  onNotReject = (uid, iid) => {
    this.props.onNotReject(uid, iid, this.addNotActiveToState);
  };

  addNotActiveToState = () => {
    this.setState({ active: false });
    this.props.onAddAttendee();
  };

  render() {
    if (this.props.isAttendeesLoading === true) {
      return (
        <div>
          <p
            style={{ display: "block", margin: "0 auto" }}
            className="lds-dual-ring"
          />
        </div>
      );
    }
    if (this.props.isAttendeesErrored === true) {
      return <p>Can not load Attendees</p>;
    }
    const {
      title,
      description,
      pictureUrl,
      createdAt,
      userId,
    } = this.props.item;

    const { classes, attendeesIds } = this.props;
    const id = this.props.getId();
    return (
      <main className={classes.main}>
        <Paper className={classes.padding + " " + classes.paper}>
          <div>
            <IconButton onClick={this.onRedirect} aria-label="Info">
              <KeyboardBackspace />
            </IconButton>
            <span>back</span>
            <Typography component="h1" variant="h5">
              Event detailed info:
            </Typography>
            <List>
              <ListItem>
                <b>Event title: &nbsp;</b> {title}
              </ListItem>
              <ListItem>
                <b>Event Description:&nbsp;</b> {description}
              </ListItem>
              <ListItem>
                <b>Event Image:&nbsp;</b>
              </ListItem>
              <ListItem>
                <img
                  className={classes.image}
                  src={pictureUrl}
                  alt={description}
                />
              </ListItem>
              <ListItem>
                <b>Event created at:&nbsp; </b>{" "}
                <Moment format="YYYY/MM/DD HH:mm:ss">{createdAt}</Moment>
              </ListItem>
              <ListItem>
                <b>Event created by:&nbsp; </b>{" "}
                <Link
                  data-user-id={userId}
                  onClick={() => this.onView(userId)}
                  key={userId}
                >
                  {" "}
                  {this.props.attendeeFormatted[userId]}{" "}
                </Link>
              </ListItem>
              <ListItem>
                <b>Users assigned to event:&nbsp;</b>
              </ListItem>
              <ListItem>
                {attendeesIds.length > 0 && (
                  <div>
                    {attendeesIds.map((attenId) => (
                      <Link
                        data-user-id={attenId}
                        onClick={() => this.onView(attenId)}
                        key={attenId}
                      >
                        {" "}
                        {this.props.attendeeFormatted[attenId]}{" "}
                      </Link>
                    ))}
                  </div>
                )}
              </ListItem>
              {attendeesIds.includes(this.props.authUser.uid) && (
                <ListItem>
                  <Button
                    variant="contained"
                    color="primary"
                    className={classes.button}
                    onClick={() =>
                      this.onNotReject(this.props.authUser.uid, id)
                    }
                    disabled={!this.state.active}
                  >
                    Accept
                  </Button>
                  <Button
                    variant="contained"
                    color="secondary"
                    onClick={() => this.onReject(this.props.authUser.uid, id)}
                    disabled={this.state.active}
                  >
                    Reject
                  </Button>
                </ListItem>
              )}
            </List>
          </div>
        </Paper>
      </main>
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  getId: () => {
    return getId(ownProps);
  },
  item: getItem(state, ownProps.match.params.id, getItems(state)),
  attendeeFormatted: getAttendeeFormatted(state, getUsersKey(state)),
  attendeesIds: getAttendeesIds(state),
  isAttendeesErrored: getAttendeesHasErrored(state),
  isAttendeesLoading: getAttendeesIsLoading(state),
  authUser: getAuthUser(state),
});

const mapDispatchToProps = (dispatch) => ({
  fetchAttendeesIds: (id) => dispatch(attendeesIdsFetchData(id)),
  onItemsOff: () => dispatch(itemsOff()),
  onUsersOff: () => dispatch(usersOff()),
  onUsersEnrolmentsListOff: () => dispatch(usersEnrolmentsListOff()),
  onReject: (uid, iid, saveActiveToStateCallback) =>
    dispatch(reject(uid, iid, saveActiveToStateCallback)),
  onNotReject: (uid, iid, saveNotActiveToStateCallback) =>
    dispatch(notReject(uid, iid, saveNotActiveToStateCallback)),
  onDeleteAttendee: () => dispatch(updateItemsInState()),
  onAddAttendee: () => dispatch(updateItemsInState()),
});

export default withStyles(styles)(
  compose(
    withRouter,
    connect(mapStateToProps, mapDispatchToProps)
  )(DetailedItem)
);
