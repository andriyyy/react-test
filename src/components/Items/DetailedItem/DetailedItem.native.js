import React, { Component } from "react";
import {
  Colors,
  Button,
  ActivityIndicator,
  Subheading,
  Surface,
  Chip,
} from "react-native-paper";
import { View, Text, Image, ScrollView, StyleSheet } from "react-native";
import { connect } from "react-redux";
import moment from "moment";

import {
  getAttendeesIds,
  getAttendeesHasErrored,
  getAttendeesIsLoading,
  getItem,
  getItems,
  getUsersKey,
  getAttendeeFormatted,
  getAuthUser,
} from "../../../selectors/Selectors";
import { attendeesIdsFetchData } from "../../../actions/users";
import {
  itemsOff,
  usersOff,
  usersEnrolmentsListOff,
  reject,
  notReject,
  updateItemsInState,
} from "../../../actions/firebase";

class DetailedItem extends Component {
  constructor(props) {
    super(props);
    this.state = {
      active: false,
    };
  }
  componentDidMount() {
    const id = this.props.route.params.id;
    this.props.fetchAttendeesIds(id);
  }

  componentWillUnmount() {
    this.props.onItemsOff();
    this.props.onUsersOff();
    this.props.onUsersEnrolmentsListOff();
  }

  onView = (id, navigation) => {
    navigation.navigate("User", { id: id });
  };

  onReject = (uid, iid) => {
    this.props.onReject(uid, iid, this.addActiveToState);
  };

  addActiveToState = () => {
    this.setState({ active: true });
    this.props.onDeleteAttendee();
  };

  onNotReject = (uid, iid) => {
    this.props.onNotReject(uid, iid, this.addNotActiveToState);
  };

  addNotActiveToState = () => {
    this.setState({ active: false });
    this.props.onAddAttendee();
  };

  render() {
    if (this.props.isAttendeesLoading === true) {
      return <ActivityIndicator animating={true} color={Colors.red800} />;
    }
    if (this.props.isAttendeesErrored === true) {
      return <Text>Can not load Attendees</Text>;
    }
    const { attendeesIds, route, navigation } = this.props;
    const id = route.params.id;

    const {
      title,
      description,
      pictureUrl,
      createdAt,
      userId,
    } = this.props.item(id);

    return (
      <ScrollView>
        <Surface style={styles.surface}>
          <View>
            <View>
              <Text>
                <Subheading>Event title: &nbsp; </Subheading>
                {title}
              </Text>
            </View>
            <View>
              <Text>
                <Subheading>Event Description:&nbsp;</Subheading>
                {description}
              </Text>
            </View>
            <View>
              <Subheading>Event Image:&nbsp;</Subheading>
            </View>
            <View>
              <Image
                source={{ uri: pictureUrl }}
                style={{ width: 120, height: 120, resizeMode: "contain" }}
              />
            </View>
            <View>
              <Text>
                <Subheading>Event created at:&nbsp;</Subheading>
              </Text>
              <Text>{moment(createdAt).format("YYYY/MM/DD HH:mm:ss")}</Text>
            </View>
            <View>
              <Text>
                <Subheading>Event created by:&nbsp;</Subheading>
              </Text>
              <View style={styles.chipOver}>
                <Chip
                  icon="information"
                  onPress={() => this.onView(userId, navigation)}
                >
                  {this.props.attendeeFormatted[userId]}
                </Chip>
              </View>
            </View>
            <View>
              <Text>
                <Subheading>Users assigned to event:&nbsp;</Subheading>
              </Text>
            </View>
            <View>
              {attendeesIds.length > 0 && (
                <View style={styles.chipOver}>
                  {attendeesIds.map((attenId) => (
                    <Chip
                      key={attenId}
                      style={styles.chip}
                      icon="information"
                      onPress={() => this.onView(attenId, navigation)}
                    >
                      {this.props.attendeeFormatted[attenId]}
                    </Chip>
                  ))}
                </View>
              )}
            </View>
            {attendeesIds.includes(this.props.authUser.uid) && (
              <View style={styles.buttonOver}>
                <Button
                  style={styles.button}
                  mode="contained"
                  onPress={() => this.onNotReject(this.props.authUser.uid, id)}
                  disabled={!this.state.active}
                >
                  Accept
                </Button>
                <Button
                  style={styles.button}
                  mode="contained"
                  onPress={() => {
                    this.onReject(this.props.authUser.uid, id);
                  }}
                  disabled={this.state.active}
                >
                  Reject
                </Button>
              </View>
            )}
          </View>
        </Surface>
      </ScrollView>
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  item: (id) => {
    return getItem(state, id, getItems(state));
  },
  attendeeFormatted: getAttendeeFormatted(state, getUsersKey(state)),
  attendeesIds: getAttendeesIds(state),
  isAttendeesErrored: getAttendeesHasErrored(state),
  isAttendeesLoading: getAttendeesIsLoading(state),
  authUser: getAuthUser(state),
});

const mapDispatchToProps = (dispatch) => ({
  fetchAttendeesIds: (id) => dispatch(attendeesIdsFetchData(id)),
  onItemsOff: () => dispatch(itemsOff()),
  onUsersOff: () => dispatch(usersOff()),
  onUsersEnrolmentsListOff: () => dispatch(usersEnrolmentsListOff()),
  onReject: (uid, iid, saveActiveToStateCallback) =>
    dispatch(reject(uid, iid, saveActiveToStateCallback)),
  onNotReject: (uid, iid, saveNotActiveToStateCallback) =>
    dispatch(notReject(uid, iid, saveNotActiveToStateCallback)),
  onDeleteAttendee: () => dispatch(updateItemsInState()),
  onAddAttendee: () => dispatch(updateItemsInState()),
});

const styles = StyleSheet.create({
  surface: {
    padding: 15,
    margin: 20,
    alignItems: "center",
    justifyContent: "flex-start",
    elevation: 4,
  },
  buttonOver: {
    flexDirection: "row",
    justifyContent: "space-between",
  },
  button: {
    margin: 10,
  },
  chipOver: {
    flexDirection: "row",
    alignItems: "flex-start",
    flexWrap: "wrap",
  },
  chip: { margin: 4 },
});

export default connect(mapStateToProps, mapDispatchToProps)(DetailedItem);
