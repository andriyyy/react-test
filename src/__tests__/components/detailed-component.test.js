import React from "react";
import { shallow, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import DetailedItem from "../../../src/components/Items/DetailedItem/DetailedItem";


configure({ adapter: new Adapter() });

describe("<DetailedItem />", () => {
  const wrapper = shallow(<DetailedItem />);
  it("renders properly", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
