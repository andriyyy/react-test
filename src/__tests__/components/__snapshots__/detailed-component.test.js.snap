// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DetailedItem /> renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithStyles(Component) />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "classes": Object {
        "formControl": "Component-formControl-6",
        "image": "Component-image-7",
        "main": "Component-main-3",
        "margin": "Component-margin-1",
        "padding": "Component-padding-2",
        "paper": "Component-paper-4",
        "submit": "Component-submit-5",
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "classes": Object {
          "formControl": "Component-formControl-6",
          "image": "Component-image-7",
          "main": "Component-main-3",
          "margin": "Component-margin-1",
          "padding": "Component-padding-2",
          "paper": "Component-paper-4",
          "submit": "Component-submit-5",
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
